### PR TITLE
Refactor db loading, commit DB migrations and some prod changes

### LINKED ISSUE #4: Create an offline CornerStone parser
Currently, details are fetched from the website. This isn't critical, but we better support a real offline environment.

### LINKED ISSUE #52: Simplify population of HUJI DB
In preparation of deployment to the cloud(#29), there should be an automated way of filling the database

I currently have a POC at https://github.com/nmdanny/Closure/tree/simpler_data_import, which works as follows:

1. An archive containing the parsed .JSONs is hosted on the internet (currently at a GCS bucket on my account)
2. A DB migration fetches the archive and loads it to the database.
3. During deployment, `Cloud Build` will call `python manage.py migrate` after building the backend's image, which will
  run this migration if it wasn't applied to the database yet.  

The migration version is stored in the DB itself, so further re-deploys of newer containers would not call this migration again.

A minor concern is that the parsed archive is not version controlled, if we have changes to the scraper/parsed outputs, we need to remember to re-run the scraper and update the archive. This can probably be solved more elegantly via a CI/CD pipeline (e.g Cloud Build + Git triggers + Artifact Registry), though it's probably overkill at this stage.

### PR DESCRIPTION
**Pull request by nmdanny**
# Important 

After this PR is committed, upon changes to the model we should create migrations and commit them. 
This is because when we go into production, we can't afford to delete all tables and re-create all migrations every time
we change the models. 

Also, since the backend tests(currently only 1 test) run migrations which now load all course data,
they will take a longer time to startup. When developing you can uncomment the line 
`; addopts = --no-migrations` (remove the `;`) within `Closure_Test/pytest.ini`, to disable migrations and
make the test DB infer its schema from the model class metadata directly.

However, in CI we should use those migrations, as running the DB test also indirectly tests the
success of those migrations. So if you do uncomment this line, do not commit it.

# Changes

- Move functions used for loading data into `rest_api/huji_loader.py`
- Use bulk SQL operations for importing Courses and Tracks (I've yet to do this for course-groups though)
- Add function for loading corner stone status from file, closes #4
- Add functions for loading parsed data from `.zip` locally or from the internet
- Factor some constant paths located within the various parser files into variables
- Add DB migration that loads parsed data from a GCP bucket, closes #52
  (simplified README accordingly)
- Add DB migration that creates a superuser(useful for prod)
- Model change: Use TextFields instead of CharFields
  
  The reasoning is: Both PostgreSQL and SQLite can handle arbitrary length text
  fields well, moreover, SQLite doesn't even enforce varchar length constraints, in fact
  there are many entries that violate these constraints (which caused PGSQL in production to
  fail loading data)

  From a security standpoint, these fields were used for texts that were obtained from HUJI's site
  and manually vetted, and not arbitrary user input which might cause denial of service on the DB.
- Generate and commit DB migrations for the model

- Stop doing DB migrations in the cloudbuild file. The network link between CloudBuild instances and
  Cloud SQL proxy is extremely slow, posing a severe bottleneck
  for filling the data. Currently, this will be done manually.

- Modify model test to work on non-existing courses, otherwise there'll be a conflict with data
  loaded from the migration

- Some other prod changes (frontend .env file changes, cloudbuild changes and make
  Django aware of a HTTPS->HTTP proxy)

### CODE REVIEW

File: Closure_Project/rest_api/huji_loader.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+def import_course_group(group_values: Dict) -> None:
+    track_id = group_values['track_id']
+    track = Track.objects.get(track_number=track_id)
+    del group_values['track_id']
+
+    group_values['track'] = track
+
+    course_ids = group_values['course_ids']
+    del group_values['course_ids']
```
Comment (ScDor) [inline Closure_Project/rest_api/huji_loader.py:41]: `course_ids = group_values.pop('course_ids')` would be prettier, but doesn't really matter

File: README.md [hunk 2]
```diff
@@ -48,10 +48,9 @@ structures used.
 <a name="HUJIex"/>
 
 #### Loading data offline ####
-it's recommended to load the data of all courses and tracks, rather than download and re-parse thousands of files. 
-1. Download the [7z dump file](https://drive.google.com/file/d/1TQSQ--VWFKt0CCZlBOZRUxKuH6Y6u0DL/view?usp=sharing) (HUJI account required)
-2. Extract it into the `Parser` folder
-3. Run `OfflineParser.py -> load_all_dumped()` _(Tip: if fetching corner-stone coursers gets stuck, comment out that line and continue with out that data, there's an [open issue](https://github.com/ScDor/Closure/issues/4))_
+
+Upon migrating the database, a data dump containing all parsed course and track data will be
+downloaded from the internet and inserted into the database.
```
Comment (ScDor) [inline README.md:53]: Say someone forks this, will they download the data from our host as well?

File: Closure_Project/rest_api/migrations/0001_initial.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
+# Generated by Django 3.2.4 on 2021-06-27 14:11
```
Comment (ScDor) [inline Closure_Project/rest_api/migrations/0001_initial.py:1]: Are migration files safe to commit? (bug-wise, not security-wise)

File: README.md (hunk 2 above)
Comment (ScDor) [reply README.md:53]: just saw it's from Github, never mind

**Review by ScDor — APPROVED** [review]

File: README.md [hunk 4]
```diff
@@ -48,10 +48,9 @@ structures used.
 <a name="HUJIex"/>
 
 #### Loading data offline ####
```
Comment (ScDor) [inline README.md:50]: > they will take a longer time to startup. When developing you can uncomment the line
; addopts = --no-migrations (remove the ;) within Closure_Test/pytest.ini, to disable migrations and
make the test DB infer its schema from the model class metadata directly.

Please mention this on `README.md`, so future devs know it?

File: frontend-cloudbuild.yaml [hunk 5]
```diff
@@ -81,8 +81,8 @@ steps:
       ]
 
 substitutions:
-  _IMAGE_NAME: closure-service-frontend
-  _FRONTEND_BUCKET_NAME: closuretest.duckdns.org
+  _IMAGE_NAME: closure-frontend
+  _FRONTEND_BUCKET_NAME: uniclosure.me
 
 images:
   - "gcr.io/${PROJECT_ID}/${_IMAGE_NAME}:$COMMIT_SHA"
```
Comment (ilanerukh) [inline frontend-cloudbuild.yaml:88]: Should we set the COMMIT_SHA variable before executing the script?

File: backend-cloudbuild.yaml [hunk 6]
```diff
@@ -94,11 +95,12 @@ steps:
...
-  _INSTANCE_NAME: closure-postgres
+  _INSTANCE_NAME: prod-db
   _REGION: europe-west1
-  _SERVICE_NAME: closure-service
+  _SERVICE_NAME: prod-api
   _SECRET_SETTINGS_NAME: django_settings
 
 images:
   - "gcr.io/$PROJECT_ID/${_SERVICE_NAME}:$COMMIT_SHA"
+
```
Comment (ilanerukh) [inline backend-cloudbuild.yaml:105]: Should we set the COMMIT_SHA variable before executing the script?

File: frontend-cloudbuild.yaml (hunk 5 above)
Comment (nmdanny) [reply frontend-cloudbuild.yaml:88]: No, it is a magic variable automatically set by Cloud Build. Though I believe you would have to substitute it manually if you want to manually push a build, not via git.

File: Closure_Project/rest_api/migrations/0001_initial.py (hunk 3 above)
Comment (nmdanny) [reply Closure_Project/rest_api/migrations/0001_initial.py:1]: Generally yes, migration files are like any other code, at any point in time the migrations should match up to the models  (such that if you clone a repository, doing `makemirations rest_api` should not create any new files, in fact maybe I'll even make a github action for this)

File: README.md (hunk 4 above)
Comment (nmdanny) [reply README.md:50]: Done

File: README.md (hunk 2 above)
Comment (nmdanny) [reply README.md:53]: Actually, although I've attached a zip-file to the repo, the migration currently loads the zip from a public GCP bucket.
Do you have any opinions on which method is preferable? 

As I see it:

- on the one hand, a zipfile in the repo is version controlled
- on the other hand, one day we might want to decouple the parser and the parsed artifacts from the backend repo, and in that case we'd need to download it remotely from some server anyway

**Comment by ScDor** [comment]
VC plus some simplicity give github a better stand, imo

File: Closure_Project/rest_api/migrations/0001_initial.py (hunk 3 above)
Comment (ScDor) [reply Closure_Project/rest_api/migrations/0001_initial.py:1]: Perfect, thanks

File: README.md (hunk 2 above)
Comment (ScDor) [reply README.md:53]: I see version control and the relative simplicity as a slight advantage of github. Should we want to change it later (decoupling etc), it shouldn't be a big issue anyway.

Comment (nmdanny) [reply README.md:53]: Understandable, I'll switch it to use the zip file

**Comment by nmdanny** [comment]
Modified backend workflow to ensure all needed migrations were committed(and realized I missed a migration after the merge :) )

File: frontend-cloudbuild.yaml (hunk 5 above)
Comment (ilanerukh) [reply frontend-cloudbuild.yaml:88]: I think we should add it in the docs of deployment (building and pushing manually a build) to set this variable.

Comment (nmdanny) [reply frontend-cloudbuild.yaml:88]: Agreed, I also added a section for the frontend

**Comment by ScDor** [comment]
Fantastic work, merge away :)

**Pull request merged**